src/main/java/net/neoforged/neoforge/transfer/storage/StorageWithSimulations.java

```diff
@@ -0,0 +1,184 @@
+        }
+
+        return amount;
+    }
+
+    /**
+     * @param index The index to get the resource from.
+     * @return The resource at the given index.
+     */
+    // TODO: rename index to slot?
+    T getResource(int index);
+
+    /**
+     * @param index The index to get the amount from.
+     * @return The amount of the resource at the given index.
+     */
+    int getAmount(int index);
+
+    /**
+     * Gets the maximum amount that the given index can have of the given resource. If your capacity is constant, no matter
+     * the resource, you can just return the result of {@link #getCapacity(int)}. This is historically the case for fluids,
+     * but not for items.
+     *
+     * @param index The index to get the limit from.
+     * @param resource The resource to get the limit for.
+     * @return The limit of the resource at the given index.
+     */
+    int getCapacity(int index, T resource);
+
+    /**
+     * Gets the theoretical maximum amount that the given index can hold of a resource.
+     *
+     * @param index The index to get the limit from.
+     * @return The limit of the resource at the given index.
+     */
+    // TODO: just pass blank resource, this is redundant
+    int getCapacity(int index);
+
+    /**
+     * Checks if the given resource is allowed to be inserted into the handler at the given index, regardless of the
+     * current state of the handler.
+     *
+     * @param index The index to check.
+     * @param resource The resource to check.
+     * @return True if the resource can be inserted, false otherwise.
+     */
+    boolean isValid(int index, T resource);
+
+    /**
+     * Checks if the given index allows insertion of a resource, regardless of the state of the handler.
+     *
+     * <p>
+     * As long as the handler could, under the right conditions, allow a resource to be inserted into the given index,
+     * this should return true.
+     *
+     * @param index The index to check.
+     * @return True if the resource can be inserted, false otherwise.
+     */
+    // TODO: too fine-grained IMO
+    default boolean allowsInsertion(int index) { return true; }
+
+    /**
+     * Checks if the given index allows extraction of a resource, regardless of the state of the handler.
+     * <p>
+     * As long as the handler could, under the right conditions, allow a resource to be extracted from the given index,
+     * this should return true.
+     *
+     * @param index The index to check.
+     * @return True if the resource can be extracted, false otherwise.
+     */
+    // TODO: too fine-grained IMO
+    default boolean allowsExtraction(int index) { return true; }
+
+    /**
+     * Checks if the handler allows insertion into at least one index, regardless of the state of the handler.
+     *
+     * @return True if a resource can be inserted, false otherwise.
+     */
+    default boolean allowsInsertion() {
+        for (int i = 0; i < size(); i++) {
```

Comment:
We should store the size instead of query it each loop iteration.

Reply:
Per-slot `allowsInsertion` was removed so I am marking this as resolved.